src/main.zig

```diff
@@ -2240,6 +2251,49 @@ fn buildOutputType(
                     },
+                    .sanitize_trap, .no_sanitize_trap => |t| {
+                        const enable = t == .sanitize_trap;
+                        var san_it = std.mem.splitScalar(u8, it.only_arg, ',');
+                        var recognized_any = false;
+                        while (san_it.next()) |sub_arg| {
+                            // This logic doesn't match Clang 1:1, but it's probably good enough, and avoids
+                            // significantly complicating the resolution of the options.
+                            if (mem.eql(u8, sub_arg, "undefined")) {
+                                if (mod_opts.sanitize_c) |sc| switch (sc) {
+                                    .off => if (enable) {
+                                        mod_opts.sanitize_c = .trap;
+                                    },
+                                    .trap => if (!enable) {
+                                        mod_opts.sanitize_c = .full;
+                                    },
+                                    .full => if (enable) {
+                                        mod_opts.sanitize_c = .trap;
+                                    },
+                                } else {
+                                    if (enable) {
+                                        mod_opts.sanitize_c = .trap;
+                                    } else {
+                                        // This means we were passed `-fno-sanitize-trap=undefined` and nothing else. In
+                                        // this case, ideally, we should use whatever value `sanitize_c` resolves to by
+                                        // default, except change `trap` to `full`. However, we don't yet know what
+                                        // `sanitize_c` will resolve to! So we either have to pick `off` or `full`.
+                                        //
+                                        // `full` has the potential to be problematic if `optimize_mode` turns out to
+                                        // be `ReleaseFast`/`ReleaseSmall` because the user will get a slower and larger
+                                        // binary than expected. On the other hand, if `optimize_mode` turns out to be
+                                        // `Debug`/`ReleaseSafe`, `off` would mean UBSan would unexpectedly be disabled.
+                                        //
+                                        // `off` seems very slightly less bad, so let's go with that.
+                                        mod_opts.sanitize_c = .off;
+                                    }
+                                }
+                                recognized_any = true;
+                            }
+                        }
+                        if (!recognized_any) {
+                            try cc_argv.appendSlice(arena, it.other_args);
+                        }
+                    },
```

Comment:
I kind of hate this, but I'm not sure what to do about it.

I considered splitting `sanitize_c` into two fields: `sanitize_c` and `sanitize_c_prefer_trap`. But that just runs into the problem that we can't assign a meaningful value here:

https://github.com/ziglang/zig/blob/9d6b20fff06da9b34c1d5bf2ce383004b0c70644/src/main.zig#L2948-L2954

(Because the user might only have provided one or the other explicitly, and we need both to answer the question of whether any module in the compilation needs the UBSan runtime.)

It almost seems like iterating the full module graph here to determine whether the UBSan runtime is needed would be easier:

https://github.com/ziglang/zig/blob/9d6b20fff06da9b34c1d5bf2ce383004b0c70644/src/Compilation.zig#L1290-L1297

Anyway, ideas welcome.